src/journal-remote/header-util.c

```diff
@@ -0,0 +1,170 @@
+                       ","       \
+                       "\\"      \
+                       "/"       \
+                       "'"       \
+                       "\""      \
+                       "?"       \
+                       "!"       \
+                       "("       \
+                       ")"       \
+                       "{"       \
+                       "}"       \
+                       "["       \
+                       "]"       \
+                       "@"       \
+                       "<"       \
+                       ">"       \
+                       "="       \
+                       "-"       \
+                       "+"       \
+                       "*"       \
+                       "#"       \
+                       "$"       \
+                       "&"       \
+                       "`"       \
+                       "|"       \
+                       "~"       \
+                       "^"       \
+                       "%"
+
+static bool header_name_is_valid(const char *e, size_t n) {
+        if (!e)
+                return false;
+
+        if (n > VALID_HEADER_NAME_LENGTH)
+                return false;
+
+        for (const char *p = e; p < e + n; p++)
+                if (!strchr(VALID_HEADER_NAME_CHARS, *p))
+                        return false;
+
+        return true;
+}
+
+static bool header_value_is_valid(const char *e) {
+        if (!e)
+                return false;
+
+        int n = strlen_ptr(e);
+
+        if (n < 0)
+                return false;
+
+        for (const char *p = e; p < e + n; p++)
+                if (!strchr(VALID_HEADER_VALUE_CHARS, *p))
+                        return false;
+
+        return true;
+}
+
+bool header_is_valid(const char *e) {
+        const char *eq;
+
+        eq = strchr(e, ':');
+        if (!eq)
+                return false;
+
+        if (!header_name_is_valid(e, eq - e))
+                return false;
+
+        if (!header_value_is_valid(skip_leading_chars(eq + 1, WHITESPACE)))
+                return false;
+
+        return true;
+}
+
+static bool header_entry_has_name(const char *entry, const char *name) {
+        const char *t;
+
+        assert(entry);
+        assert(name);
+
+        t = startswith(entry, name);
+        if (!t)
+                return false;
+
+        return *t == ':';
+}
+
+static char **strv_header_unset(char **l, const char *name) {
+        assert(name);
+
+        if (!l)
+                return NULL;
+
+        /* Drops every occurrence of the header var setting p in the
+         * string list. Edits in-place. */
+
+        char **f, **t;
+        for (f = t = l; *f; f++) {
+                if (header_entry_has_name(*f, name)) {
+                        free(*f);
+                        continue;
+                }
+
+                *(t++) = *f;
+        }
+
+        *t = NULL;
+        return l;
+}
+
+int strv_header_replace_consume(char ***l, char *p) {
```

Comment:
i am really not convinced we should erase earlier entries with the same header field name. i think we should just append to the end, accepting duplicates

Reply:
you are right, updated implementation